### ISSUE TITLE
IPC deprecation warning appears on OSX, but not Windows

### ISSUE THREAD

**Issue by Slapbox** [electron/electron#22754]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:** 8.1.0
* **Operating System:** macOS 10.13.6
* **Last Known Working Electron version:** N/A

### Expected Behavior
Sending a function over IPC should show deprecation warning on any OS.

> DeprecationWarning: Passing functions, DOM objects and other non-cloneable JavaScript objects to IPC methods is deprecated and will throw an exception beginning with Electron 9.

### Actual Behavior
Warning only appears on macOS.

### To Reproduce
```js
ipcRenderer.send('some-channel', () => {});
```

**Comment by codebytere** [comment]
cc @nornagon

**Comment by nornagon** [comment]
pretty sure this is just because of `stdout` being weird on windows. I don't think it's specific to this message.

**Comment by Slapbox** [comment]
Just seemed weird because other deprecation warnings have always appeared for me. Perhaps I've simply never seen a deprecation warning from the `main` process before.

This causes us no issues here, I just wanted to report it. I'm going to leave it to the Electron team to determine whether this should be closed.

**Comment by codebytere** [comment]
Closing as I think this is working as expected, it's just somewhat unintuitive owing to, as Jeremy said above, weirdness with `stdout` on Windows systems.

**Comment by Slapbox** [comment]
It seems like not every situation that should create a deprecation warning creates one even on macOS. 

We just tried to upgrade to `Electron@9.3.x` but we get the message `Error: An object could not be cloned` - So I headed to macOS to see what I'd missed, but I can't provoke any deprecation warnings to appear.

I thought we'd resolved all situations where we were passing non-serializable data over IPC, but I guess not. Unfortunately this error has no stack trace or anything else.

Without version 8 showing deprecation warnings in every case where it should this makes for rough debugging.

Is there a chance that something that should warn isn't warning as intended?

**Comment by Slapbox** [comment]
I reviewed all the invocations of `webContents.send`, `ipcMain` and `ipcRenderer` in our app and I don't see anything that I'd expect to trigger this issue. I commented out virtually everything but the issue persists. 

Any ideas would be *greatly* appreciated.

**Comment by nornagon** [comment]
@Slapbox do you have a repro using [Fiddle](/electron/fiddle)?

**Comment by Slapbox** [comment]
@nornagon thanks for your reply. Unfortunately I don't. I can look into creating one. The problem is that right now I really don't know what could be triggering it as I don't believe we're sending anything but serializable objects and plain strings.

Is data that travels via `remote` subject to the serializability rules? It seems not based on my research and testing, but if that's the case then I'm really at a loss.

Was there any change in IPC between versions 8 & 9 besides the transition from deprecation to throwing and error? I ask because it's listed as a breaking change in 8 & 9, but I suspect that's the extent of the changes?

**Comment by nornagon** [comment]
@Slapbox yes, remote uses the same serialization system as all other IPC.

**Comment by Slapbox** [comment]
@nornagon thanks for that info!

That was my initial suspicion, but when I was playing with Fiddle I had thought I managed to send `BrowserWindow` over `remote`, which got me thinking `remote` must not use `ipc`. I'll re-review everything now with that in mind. Thanks again, I really appreciate it!

**Comment by Slapbox** [comment]
@nornagon Fiddles definitely allow `remote` to transmit/receive these objects unless something it really bugged on my machine or in my head https://gist.github.com/2edf03080ace043e34e96af4ddd9b609

**Comment by Slapbox** [comment]
We definitely are not seeing deprecation warnings on Linux with Electron@8.5.5. I've tried to provoke them by playing with any remaining uses of `remote` that we have, but I can't produce any warnings.

Upon upgrading to Electron@9.x though, we get errors that we're sending non-clonable objects.

I'd really like to be more helpful in providing info on what code is not producing the deprecation warnings, but unfortunately I can't reproduce any of this in Fiddles because anything sends over `remote` without issue.

Can anyone give us any advice on this? As I understand it Electron 8 is no longer supported and we've been trying to upgrade, but this is making upgrading pretty much impossible.

**The issue with not producing warnings must be related to `remote.getGlobal()` because we're definitely not sending anything non-clonable via traditional IPC.**

**Comment by Slapbox** [comment]
Alright I guess I don't understand what can and cannot be sent over `remote.getGlobal` - seems like anything at all is allowed. I find the docs on this really confusing and lacking. I'd happily offer to improve them if I had the knowledge.

**I did manage to fix my issue.** For anyone else who runs into these mystery IPC warnings upon upgrading to Electron@9.x, the errors don't seem to be related to `IPC` or `remote` usage at all. Electron didn't like something in our preload script and failed to throw a relevant error and instead threw that. I have guesses about why, but not educated ones so I'll keep them to myself.

Wrapping our preload script's logic with `process.once('document-start', () => { ... })` allowed us to debug the preload script and test our code, and by removing that sort of blackbox we were able to help ourselves and get things working again.

**Comment by Slapbox** [comment]
Still getting these errors with no possible way to determine what the source is. Electron@8.x shows nothing. We upgrade and suddenly we get `An object could not be cloned.`

**Any advice for how to find the source of these would be _greatly_ appreciated.**

Between no deprecation warnings and no callstacks on the error, it's an absolute black box.

**Comment by nornagon** [comment]
@Slapbox it's been a while since I looked at this issue, was there ever an isolated repro showing the issue?

Does the debugger help? Can you attach the inspector and break on all exceptions? That might give you a call stack.

**Comment by Slapbox** [comment]
@nornagon thanks for your reply.

Eventually I found that it seems to be caused by using `electron-log`, specifically version `3.x`. Upgrading did resolve it.

Whatever about the logger was triggering the errors - it didn't produce warnings in `8.x` even on Linux/macOS.

I tried for a long time to debug the issue directly like your suggestion but no call stacks or anything helpful no matter how much I searched for some place to get a foot in the door and find a way to debug this. 

In retrospect, the fact it was our logger itself producing the problem did no favors for finding the resolution.

While there's definitely a real issue in here with warnings not triggering somehow, I doubt it's worth looking into any further unless someone else runs into the same issue.

Thanks again very much for your reply.
